File: src/main/java/com/myapp/auth/TokenAuthenticationServiceImpl.java
package com.myapp.auth;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.security.core.Authentication;
import org.springframework.stereotype.Service;

import javax.servlet.http.HttpServletRequest;

@Service
class TokenAuthenticationServiceImpl implements TokenAuthenticationService {

    private final TokenHandler tokenHandler;

    @Autowired
    TokenAuthenticationServiceImpl(TokenHandler tokenHandler) {
        this.tokenHandler = tokenHandler;
    }

    public Authentication getAuthentication(HttpServletRequest request) {
        final String authHeader = request.getHeader("authorization");
        if (authHeader == null) return null;
        if (!authHeader.startsWith("Bearer")) return null;

        final String jwt = authHeader.substring(7);
        if (jwt.isEmpty()) return null;

        return tokenHandler
                .parseUserFromToken(jwt)
                .map(UserAuthentication::new)
                .orElse(null);
    }
}

File: src/main/kotlin/com/myapp/auth/TokenAuthenticationServiceImpl.kt
package com.myapp.auth

import org.springframework.security.core.Authentication
import org.springframework.stereotype.Service
import javax.servlet.http.HttpServletRequest

@Service
class TokenAuthenticationServiceImpl(
    private val tokenHandler: TokenHandler
) : TokenAuthenticationService {

    override fun authentication(request: HttpServletRequest): Authentication? {
        val authHeader = request.getHeader("authorization") ?: return null
        if (!authHeader.startsWith("Bearer")) return null

        val jwt = authHeader.substring(7)
        if (jwt.isEmpty()) return null

        return tokenHandler
            .parseUserFromToken(jwt)
            .let(::UserAuthentication)
    }
}

